crucible-llvm/src/Lang/Crucible/LLVM/MemModel/Partial.hs

```diff
@@ -318,21 +319,25 @@ totalLLVMVal :: (IsExprBuilder sym)
 totalLLVMVal sym = NoErr (truePred sym)
 
 -- | Take a partial value and assert its safety
-assertSafe :: (IsSymInterface sym)
+assertSafe :: (IsSymInterface sym, ?memOpts :: MemOptions)
```

Comment:
One thing that just dawned on me: I'm modifying all uses of `assertSafe`, but really, I think we only care about the particular call site in `readMem'`, which is responsible for checking reads of unwritten memory. Perhaps I should create a separate version of `assertSafe` for this specific need?

Reply:
I think that would be a sensible thing to do, yeah.  At a minimum, this new operation should have a different name.

Reply:
This idea ends up being trickier to implement than I imagined. `readMem'` itself isn't actually the function that invokes `assertSafe`. Rather, it returns an `Err` value that is eventually passed to `doLoad`, which invokes `assertSafe` on it.

One idea I had was to make a specialized version of `assertSafe` that is specific to `doLoad`, but even that's too broad. There are other functions besides `readMem'` that produce `Err` values that make it to `doLoad`, such as `readMemInvalidate` (which `Err`s upon reading invalidated memory), `readMemStore` (which `Err`s upon reading invalid memory), and `readMem` (which can potentially have a failing `NoErr` due to bad alignment or an unreadable region).

I suppose if we wanted to distinguish between these cases, we'd have to check what `MemoryErrorReason` was used to annotate the `Pred` attached to the `PartLLVMVal`. Is it feasible to do so?

Reply:
I don't think looking at the `MemoryErrorReason` is a good idea.

Reply:
Maybe the right play is to go back to our original idea.  You can use the same `freshLLVMVal` you just wrote in the base case of `readMem'` whenever `readUnwrittenMemory` is enabled.

Reply:
Ah, of course. That should have been obvious to me in hindsight, but better late than never.